# swagger_client
The Gene Feature Enumeration (GFE) Submission service provides an API for converting raw sequence data to GFE. It provides both a RESTful API and a simple user interface for converting raw sequence data to GFE results. Sequences can be submitted one at a time or as a fasta file. This service uses <a href=\"https://github.com/nmdp-bioinformatics/service-feature\">nmdp-bioinformatics/service-feature</a> for encoding the raw sequence data and <a href=\"https://github.com/nmdp-bioinformatics/HSA\">nmdp-bioinformatics/HSA</a> for aligning the raw sequence data. The code is open source, and available on <a href=\"https://github.com/nmdp-bioinformatics/service-gfe-submission\">GitHub</a>.<br><br>Go to <a href=\"http://service-gfe-submission.readthedocs.io\">service-gfe-submission.readthedocs.io</a> for more information

This Python package is automatically generated by the [Swagger Codegen](https://github.com/swagger-api/swagger-codegen) project:

- API version: 1.0.7
- Package version: 1.0.0
- Build package: io.swagger.codegen.languages.PythonClientCodegen

## Requirements.

Python 2.7 and 3.4+

## Installation & Usage
### pip install

If the python package is hosted on Github, you can install directly from Github

```sh
pip install git+https://github.com//.git
```
(you may need to run `pip` with root permission: `sudo pip install git+https://github.com//.git`)

Then import the package:
```python
import swagger_client 
```

### Setuptools

Install via [Setuptools](http://pypi.python.org/pypi/setuptools).

```sh
python setup.py install --user
```
(or `sudo python setup.py install` to install the package for all users)

Then import the package:
```python
import swagger_client
```

## Getting Started

Please follow the [installation procedure](#installation--usage) and then run the following:

```python
from __future__ import print_function
import time
import swagger_client
from swagger_client.rest import ApiException
from pprint import pprint
# create an instance of the API class
api_instance = swagger_client.DefaultApi()
locus = 'locus_example' # str | Valid HLA, KIR or ABO locus (optional)
file = '/path/to/file.txt' # file | HML file (optional)
retry = 56 # int | Number of retry requests to feature-service (optional)
feature_url = 'feature_url_example' # str | URL for the feature-service (optional)
structures = true # bool | Flag for returning the sequence and accession number of each feature (optional)
verbose = true # bool | Flag for running service in verbose (optional)

try:
    api_response = api_instance.fasta_post(locus=locus, file=file, retry=retry, feature_url=feature_url, structures=structures, verbose=verbose)
    pprint(api_response)
except ApiException as e:
    print("Exception when calling DefaultApi->fasta_post: %s\n" % e)

```

## Documentation for API Endpoints

All URIs are relative to *https://localhost*

Class | Method | HTTP request | Description
------------ | ------------- | ------------- | -------------
*DefaultApi* | [**fasta_post**](docs/DefaultApi.md#fasta_post) | **POST** /fasta | 
*DefaultApi* | [**flowhml_post**](docs/DefaultApi.md#flowhml_post) | **POST** /flowhml | 
*DefaultApi* | [**gfe_post**](docs/DefaultApi.md#gfe_post) | **POST** /gfe | 
*DefaultApi* | [**hml_post**](docs/DefaultApi.md#hml_post) | **POST** /hml | 
*DefaultApi* | [**sequence_post**](docs/DefaultApi.md#sequence_post) | **POST** /sequence | 


## Documentation For Models

 - [Error](docs/Error.md)
 - [Sequence](docs/Sequence.md)
 - [Structure](docs/Structure.md)
 - [Subject](docs/Subject.md)
 - [SubjectData](docs/SubjectData.md)
 - [Typing](docs/Typing.md)
 - [TypingData](docs/TypingData.md)


## Documentation For Authorization

 All endpoints do not require authorization.


## Author

mhalagan@nmdp.org

