Overview
========

GFE is described in more detail on the `IGDAWG website`_.

.. _feature service: http://feature.nmdp-bioinformatics.org/
.. _A gene feature enumeration approach for describing HLA allele polymorphism: http://dx.doi.org/10.1016/j.humimm.2015.09.016
.. _IGDAWG website: http://igdawg.org/enumeration.html

Paper
------------------------
In 2015 the paper introducing GFE ("`A gene feature enumeration approach for describing HLA allele polymorphism`_") was published by Steve Mack in Human Immunology.
The paper describes the GFE notation, its applicaitons, and the development of the feature service. 


GFE Nomenclature
------------------------
.. note:: Currently, the GFE nomenclature returned from the service does not represent partial sequences. In the next release we intend to represent this as an *aligned* percent for each *feature* to the reference sequence. This is similar to the *aligned* value returned for the full sequence.

The GFE allele name is created by joining the *feature* accesion numbers.
Each accession number is generated by making a RESTful call to the `feature service`_ with the sequence and it's corresponding *feature*.
If the sequence has been observed before, the the `feature service`_ will return the previously generated accession number, otherwise it will generate a unique accession number for that feature and return it.
The GFE allele **HLA-Aw1-1-7-20-10-32-7-1-1-1-6-1-5-3-5-1-1** is broken up into the following parts:

+-----------------+--------------------+
| Feature         | Accession Number   |
+-----------------+--------------------+
| five_prime_UTR  | 1                  |
+-----------------+--------------------+
| Exon 1          | 1                  |
+-----------------+--------------------+
| Intron 1        | 7                  |
+-----------------+--------------------+
| Exon 2          | 20                 |
+-----------------+--------------------+
| Intron 2        | 10                 |
+-----------------+--------------------+
| Exon 3          | 32                 |
+-----------------+--------------------+
| Intron 3        | 7                  |
+-----------------+--------------------+
| Exon 4          | 1                  |
+-----------------+--------------------+
| Intron 4        | 1                  |
+-----------------+--------------------+
| Exon 5          | 1                  |
+-----------------+--------------------+
| Intron 5        | 6                  |
+-----------------+--------------------+
| Exon 6          | 1                  |
+-----------------+--------------------+
| Intron 6        | 5                  |
+-----------------+--------------------+
| Exon 7          | 3                  |
+-----------------+--------------------+
| Intron 7        | 5                  |
+-----------------+--------------------+
| Exon 8          | 1                  |
+-----------------+--------------------+
| three_prime_UTR | 1                  |
+-----------------+--------------------+

In the example above all of the features for HLA-A were found in the sequence.
If a sequence does not contain a particular feature, then the feature accession number will be zero. 
The raw sequence is broken up into each of its features using the annotation pipeline.


